### ISSUE TITLE
Script compilations created from a previous compilation have incorrect return type

### ISSUE THREAD

**Issue by abock** [dotnet/roslyn#17403]
We are now finally able to use the Roslyn 2.0 NuGet for Xamarin Workbooks, which means I can finally leverage my work for [supporting compilations against a target (not host) corlib](https://github.com/dotnet/roslyn/pull/8507).

Unfortunately the tests I wrote for that fix were not extensive enough to catch an issue where subsequent compilations result in an implicit reference to host corlib and not the target corlib.

This is due to the implementation of `ScriptCompilationInfo.ReturnType`, which was desired to never return `null`:

```csharp
public Type ReturnType => ReturnTypeOpt ?? typeof(object);
```

I did not catch all the places in Roslyn that were using `ReturnType` but should now be using `ReturnTypeOpt`:

* `CSharpScriptCompilationInfo.WithPreviousScriptCompilation`
* `CSharpCompilation.WithScriptCompilationInfo`
* `VisualBasicScriptCompilationInfo.WithPreviousScriptCompilation`
* `VisualBasicCompilation.WithScriptCompilationInfo`

This is how the bug ends up manifesting in Workbooks:

<img width="628" alt="screen shot 2017-02-26 at 3 44 52 pm" src="https://cloud.githubusercontent.com/assets/49539/23343501/99d63ee6-fc3a-11e6-898d-0511b0b706bc.png">
